### DISCUSSION TITLE
Automatic Odometer Readings

### DISCUSSION THREAD

**Discussion by mikaelacaron** [mikaelacaron/Basic-Car-Maintenance#145] [Ideas]
I want to have some way to track the odometer readings.

The first step would be a simple view to enter the mileage / kilometers for a specific vehicle with a date, and then you could see a graph as time goes on.

But I think there's a couple really interesting scenarios that could happen for this feature:

### Scenario 1
* Create shortcuts that automatically log miles / km as you drive
* Start shortcut, to start "tracking," it's triggered by connecting to your car via bluetooth in the Shortcuts app
* Stop "tracking" when your phone disconnects from bluetooth
* The app automatically logs your distance driven after that
* And even further, you could see a map of where you drove

### Scenario 2
* Use shortcuts to bring up a view for taking a picture of your odometer when you first get in your car
* A different shortcut for taking a picture of your odometer after arriving at your destination
* Using machine learning / photo recognition to read the odometer readings from those pictures, and then subtracting to figure out how far a person drove

## Requirements
This would require several things being built, for both of these scenarios
* Shortcuts
   * Integration with Shortcuts / Automation triggers
* Map View
   * For creating a map with a path of where you've driven
* Firebase machine learning

# If anyone wants to work on this, I only want it to happen a single issue at a time! (and there should be discussion here before starting any work)

**Comment by thx1766** [comment]
I think this would be a lot of fun to work on!

I like the idea of the second scenario with OCR-ing a photo of the odometer reading. This is just because it seems less involved than the first scenario. That said, the first scenario seems more rewarding to accomplish, and seems like it can be broken into small stages pretty easily.

I think a first step for the first scenario would be to work on an automation that just reminds you to log the odometer reading when connecting to your vehicle.
In my mind a logical second step would be route tracking would just be displaying a map view with start and stop pins set manually, and later implementing something to track the actual route taken between the points.

Possibly these could be independent issues, and tied together later.

**Reply by mikaelacaron** [reply]
Yeah I'm definitely thinking this needs to be done piece by piece.
First thing, let's make a manual way to first track everything, before diving into this automation. I think doing the easier thing first would probably be best, we still have all month!

I'll make an issue for creating a screen to do this, and I'll assign it to you @thx1766 !

**Reply by thx1766** [reply]
Awesome, thanks @mikaelacaron !

**Comment by Jakayus** [comment]
Scenario 2 would be really cool! One of the current roadblocks to it is being limited only to the simulator, so hard to test out the live camera option.

I'm going to experiment this week and see if I can get a way to run on device - I'll put any updates or findings here!

**Reply by Jakayus** [reply]
So I have been able to run it on my phone but it'd be great for someone else to verify. What ended up working for me is toggling the developer settings after failing to build. So assuming it fails to build on device, do the following:
- Set Developer team to "None"
- Uncheck "Automatically manage signing"
- Re-check "Automatically manage signing"
- Re-set your Developer team

This seems to get Xcode back into a good state with the apple sign in features, because I can then see the bundle ID in my developer account and the "Sign in with Apple Enabled" in the Capabilities list. 

To be completely honest I am not sure why that worked, but since then I've been able to build to device pretty comfortably.

**Reply by mikaelacaron** [reply]
You shouldn't have needed to set the development team, because that is done in the `xcconfig` was that properly filled in?

**Reply by Jakayus** [reply]
I set the development team and bundle identifier per the instructions (hence I grabbed the team Id from my developer account and used a bundle identifier I made up). But even with that set in xconfig, I was getting an error in Xcode related to the Apple sign in and provisioning I think, when building to device. It was after I reset/toggled everything that it seemed to go away.

I can do another clone and send some screenshots/error printouts and have more specific details.

**Reply by mikaelacaron** [reply]
After you reset the automagically manage signing, is there any diff in git?
There would be if you "Re-set your Developer team" in the drop down, (I'm guessing), but without doing that, is there any changes in git for only unchecking and rechecking the automatically manage signing?

**Reply by Jakayus** [reply]
That's a good question! Not at my computer anymore but I'll check and report in later tonight. And with screenshots or printouts.

**Reply by Jakayus** [reply]
You were right! I must have messed something up the first time in setup because when I deleted my files and re-cloned, I had no differences when comparing to dev - no provisioning or ID type errors after setup. It also built right away on my phone.


Unfortunately I don't have recommended instructions to run on device - following your setup steps seem to "just work" for me - the bundle ID on App Store Connect lists Sign in with Apple as a feature and I'm good to go to run on device. Perhaps someone else can verify?


Note - I toggled the automatic signing on/off for reference, and it does cause differences. I placed a screenshot below, but following the setup instructions (properly) leaves no differences. I guess Xcode resets a couple of things whenever you toggle the automatic signing.

![Toggle_Automatic_Signing](https://github.com/mikaelacaron/Basic-Car-Maintenance/assets/6628293/d35b54ea-fdbc-4f36-b134-6e0f72290155)
